core/shared/src/main/scala/fs2/util/Catenable.scala

```diff
@@ -5,70 +5,134 @@ import Catenable._
 /**
  * Trivial catenable sequence. Supports O(1) append, and (amortized)
  * O(1) `uncons`, such that walking the sequence via N successive `uncons`
- * steps takes O(N). Like a difference list, conversion to a `Stream[A]`
+ * steps takes O(N). Like a difference list, conversion to a `Seq[A]`
  * takes linear time, regardless of how the sequence is built up.
  */
 sealed abstract class Catenable[+A] {
 
   /** Returns the head and tail of this catenable if non empty, none otherwise. Amortized O(1). */
-  def uncons: Option[(A, Catenable[A])] = {
-    @annotation.tailrec
-    def go(c: Catenable[A], rights: List[Catenable[A]]): Option[(A,Catenable[A])] = c match {
-      case Empty => rights match {
-        case Nil => None
-        case c :: rights => go(c, rights)
+  final def uncons: Option[(A, Catenable[A])] = {
+    var c: Catenable[A] = this
+    var rights: List[Catenable[A]] = Nil
```

Comment:
I didn't benchmark it but I figured it would be significantly slower since whatever we use here needs O(1) cons/uncons too.